src/Xrefcheck/System.hs

```diff
@@ -36,21 +45,118 @@ readingSystem = unsafePerformIO
+--
+-- It should be created via 'canonicalizePath'.
+--
+-- Currently, canonical paths have been made absolute, normalised
+-- regarding the running platform (e.g. Posix or Windows), with
+-- indirections syntactically expanded as much as possible and
+-- with no trailing path separator. All this results in a weaker
+-- version than that provided by 'System.Directory.canonicalizePath'.
+newtype CanonicalPath = UnsafeCanonicalPath
+  { unCanonicalPath :: FilePath
+  } deriving newtype (Show, Eq, Ord)
+
+canonicalizePath :: FilePath -> IO CanonicalPath
+canonicalizePath = fmap canonicalize . Directory.makeAbsolute
+  where
+    canonicalize :: FilePath -> CanonicalPath
+    canonicalize = UnsafeCanonicalPath
+      . expandIndirections
+      . FP.normalise
+      . FP.dropTrailingPathSeparator
+
+    expandIndirections :: FilePath -> FilePath
+    expandIndirections = FP.joinPath
+      . reverse
+      . expand 0
+      . reverse
+      . FP.splitDirectories
+
+    expand :: Int -> [FilePath] -> [FilePath]
+    expand acc (".." : xs) = expand (acc + 1) xs
+    expand acc ("." : xs) = expand acc xs
+    expand 0 (x : xs) = x : expand 0 xs
+    expand acc (_ : xs) = expand (acc - 1) xs
+    expand acc [] = replicate acc ".."
+
+-- | 'FilePath.takeDirectory' version for 'CanonicalPath'.
+takeDirectory :: CanonicalPath -> CanonicalPath
+takeDirectory (UnsafeCanonicalPath p) = UnsafeCanonicalPath $ FP.takeDirectory p
 
--- | Glob pattern relative to repository root. Should be created via @mkGlobPattern@
-newtype RelGlobPattern = RelGlobPattern FilePath
+-- | 'FilePath.takeExtension' version for 'CanonicalPath'.
+takeExtension :: CanonicalPath -> String
+takeExtension (UnsafeCanonicalPath p) = FP.takeExtension p
+
+-- | Get the list of directories, canonicalized, between two given paths.
+getDirsBetweenRootAndFile :: CanonicalPath -> CanonicalPath -> [CanonicalPath]
+getDirsBetweenRootAndFile (UnsafeCanonicalPath rootPath) file =
+  case stripPrefix rootPath (unCanonicalPath (takeDirectory file)) of
+    Just path -> UnsafeCanonicalPath <$> scanl (FP.</>) rootPath directories
+      where
+        directories = FP.splitDirectories $ dropWhile FP.isPathSeparator path
+    Nothing -> []
+
+-- | Get a relative 'FilePath' from the second given path (child) with
+-- repect to the first one (root).
+--
+-- It returns Nothing if child cannot be reached from root downwards
+-- in the filesystem tree.
+--
+-- The resulting `FilePath` uses POSIX path separators.
+getPosixRelativeChild :: CanonicalPath -> CanonicalPath -> Maybe FilePath
+getPosixRelativeChild (UnsafeCanonicalPath root) (UnsafeCanonicalPath child) =
+  prepare . fmap replaceSeparator <$> stripPrefix root child
+  where
+    replaceSeparator :: Char -> Char
+    replaceSeparator c
+      | FP.isPathSeparator c = FPP.pathSeparator
+      | otherwise = c
+
+    prepare :: FilePath -> FilePath
+    prepare path = case dropWhile FP.isPathSeparator path of
+      "" -> "."
+      other -> other
```

Comment:
IMO this name is confusing a bit, prepare for what? Wouldn't it be better to call it `dropPathSeparators`?

Reply:
I will give it a proper name. It also handles specially the empty case, returning `"."` instead of `""`.